src/server/game/Spells/SpellEffects.cpp

```diff
@@ -4471,26 +4472,31 @@ void Spell::EffectSkill()
     TC_LOG_DEBUG("spells", "WORLD: SkillEFFECT");
 }
 
-/* There is currently no need for this effect. We handle it in Battleground.cpp
-   If we would handle the resurrection here, the spiritguide would instantly disappear as the
-   player revives, and so we wouldn't see the spirit heal visual effect on the npc.
-   This is why we use a half sec delay between the visual effect and the resurrection itself */
 void Spell::EffectSpiritHeal()
 {
     if (effectHandleMode != SPELL_EFFECT_HANDLE_HIT_TARGET)
         return;
 
-    /*
-    if (unitTarget->GetTypeId() != TYPEID_PLAYER)
-        return;
-    if (!unitTarget->IsInWorld())
+    Battlefield* bf = sBattlefieldMgr->GetBattlefieldToZoneId(unitTarget->GetMap(), unitTarget->GetZoneId());
+    if (bf) // battlefield is handling this
         return;
 
-    //m_spellInfo->Effects[i].BasePoints; == 99 (percent?)
-    //unitTarget->ToPlayer()->setResurrect(m_caster->GetGUID(), unitTarget->GetPositionX(), unitTarget->GetPositionY(), unitTarget->GetPositionZ(), unitTarget->GetMaxHealth(), unitTarget->GetMaxPower(POWER_MANA));
-    unitTarget->ToPlayer()->ResurrectPlayer(1.0f);
-    unitTarget->ToPlayer()->SpawnCorpseBones();
-    */
+    Creature* spiritGuide = GetCaster()->ToCreature();
+    if (Player* playerTarget = unitTarget->ToPlayer())
+    {
+        if (!playerTarget->IsInWorld())
+            return;
+
+        // skip if player does not want to live
+        if (spiritGuide && playerTarget->GetSpiritHealer() != spiritGuide->GetGUID())
```

Comment:
I think filtering the targets here is not the way it's supposed to be. If a player refuses to be alive again, he shouldn't be hit by the spell at all